### ISSUE TITLE
Callstack Plugin Sometimes Returns Ridiculous Results

### ISSUE THREAD

**Issue by nathanjackson** [panda-re/panda#411]
Occasionally, we've observed callstack plugin returning ridiculous results. The callstack almost looks as if its repeating itself over and over again, but then there will be a spurious entry. When the plugin returns a callstack that doesn't make sense, it looks like maybe a mode switch or context switch has occurred, but I don't know this conclusively.

We really need this fixed as it provides context for analysis purposes.

**Comment by moyix** [comment]
Is it possible this is related to thread switches? Currently these aren't tracked, although the callstack plugin has a disabled-by-default mode where it tries to guess when a thread switch happens by looking for large jumps in the value of the stack pointer (now that we have support for getting thread IDs in OSI, maybe that would be a better way to check?).

Also, is this a new issue (i.e., since https://github.com/panda-re/panda/pull/406) or one that you've noticed previously?

**Comment by nathanjackson** [comment]
It seems possible (maybe even likely) that this is due to thread switches. The mode\context switch idea was just a theory.

We have seen this issue before and the workaround is to hook PANDA up to GDB and use its backtrace. Using GDB is fine, but its really nice to be able to dump the callstack in a plugin's output to get a better context. I don't think it has anything to do with #406.

I think using the ASID to track anything is likely to have problems. I'm pretty sure that what PANDA uses for ASID is recycled for other processes in at least Linux. I found a forum post about ASID on ARM, I would think x86 would do something similar. See: https://community.arm.com/developer/ip-products/processors/f/cortex-a-forum/5229/address-space-identifier---asid. @moyix do you know something about ASID that I don't know?

It should be possible to use a composite key of PID\TID to track the stacks, but then you introduce a dependency on OSI which may or may not be a problem depending on your use case. It would be nice to select the callstack strategy at runtime as opposed to compile time. There are a few strategies that have various pros and cons:

1. Use the base or frame pointer + sizeof(target_ulong) to traverse the stack - I think this is what GDB does. Pro: No OSI. Con: Doesn't work for binaries compiled with `-fomit-frame-pointer`, also platform specific.

2. Track call\returns and use the PID\TID combination to look up (looks like the default mode). Pro: Should be accurate even without a base pointer. Con: Requires OSI.

3. Watch for large jumps in PC as in the disabled option but use PID\TID. Pro: Should be fast - no introspection of the guest state other than watching PC. Con: This is definitely a heuristic - it seems like there could be incorrect reports, and requires OSI. We also tried the stack heuristic and sometimes it seemed to work, other times it seemed to have some incorrect data. But maybe this is was due to not using TID.

4. If you're going to depend on OSI, you might as well get the stack just by looking at the task struct in Linux. In Windows you should be able to do something similar with the thread data structures. Maybe this could be part of OsiThread? Pro: Allows you to get the stack for any process. Con: Requires OSI support.

I was really hoping someone else could pick this up :crossed_fingers:. I can't spend a lot of time on this other than a simple fix (if there is one).

**Comment by moyix** [comment]
> I think using the ASID to track anything is likely to have problems. I'm pretty sure that what PANDA uses for ASID is recycled for other processes in at least Linux. I found a forum post about ASID on ARM, I would think x86 would do something similar. See: https://community.arm.com/developer/ip-products/processors/f/cortex-a-forum/5229/address-space-identifier---asid. @moyix do you know something about ASID that I don't know?

The meaning of ASID in PANDA is not the ARM usage there. It refers to an identifier for the root of that process's page tables. It should be as stable as the process ID (it can be reused, but only after the process exits). On x86 we use CR3 for this, for ARM it's a bit more complicated. You can see that logic here:

https://github.com/panda-re/panda/blob/a1bd57be6abd81a36fe15cfd4a08464c75eb4008/panda/src/common.c#L59-L108

I don't really understand option 4 -- how would you get the call stack from the task_struct directly? I know you can get the base of the current thread's stack, but I don't think you can get the sequence of return addresses on the stack without doing some stack walking (which is unreliable in the case of FPO as you point out).

I think I'm in favor of (2); it's pretty easy to implement. I would want this to be something that's optional and turned on only when someone uses the `-os` flag, though, which I think is doable.

I think I can find some time to do this sometime in the next few weeks, but it would help if you could append an example replay and a place where this manifests (e.g., a place to break with gdb where gdb's backtrace differs from PANDA's).

**Comment by nathanjackson** [comment]
Okay, I see. The CR3 register makes sense then. Now that you mention it, I think I might need to handle the case where a process ID (or CR3) is recycled after process exit in plugins like file_taint\ida_taint2. I don't think I've ever run into that issue since most of the recordings I use get cut down to much smaller ones using scissors, but nonetheless its probably an issue given a long enough recording or enough processes starting and stopping in a recording.

I see what you're saying about option 4 too. For some reason I was thinking there might have been more information in the task struct since the base was there, but maybe not.

So I'm more or less convinced that option 2 is a valid option but I've got two concerns:

Would you ever need to "clear" the stacks when in kernel-mode? I'm thinking of a case where there is a Kernel thread - I think those get there own task struct and by implication, a stack. Maybe kernel threads get a unique TID and that would take care of the issue?

Secondly, what about starting analysis while a binary is already running? Let's say I have a long running service that I record after the process has started. In this case, I would have missed the service's main function and any calls made by main and so they wouldn't show up in the callstack. There are definitely use cases where knowing what the stack looks like all the way back to the startup is useful, so I would think we'd want to have multiple strategies that can be selected at runtime (defaulting to one or the other of course).

I think I could justifiably implement stack walking if you want to fix the other issue. I'll try to get a recording to you in the next couple of days.

**Comment by nathanjackson** [comment]
@moyix You should have a link in your e-mail to download a zip file with a recording and some other stuff that should be helpful. Thanks for looking at this, I really do appreciate it.

**Comment by nathanjackson** [comment]
By the way, it does look like a threading issue, my included source code (for the binary being executed) should be a good test case.

**Comment by moyix** [comment]
Got it – thanks!

**Comment by nathanjackson** [comment]
@moyix were you still going to have a chance to look at this? We're able to dedicate some time to it now as its starting to become a blocker. However, if you've already started, I don't want to duplicate work.

**Comment by moyix** [comment]
Sorry, I haven't had a chance – this semester has been an extremely busy one.

**Comment by moyix** [comment]
Fixed by https://github.com/panda-re/panda/pull/443, thanks @LauraLMann!
